portfolio/src/main/java/com/google/sps/servlets/ProjectsServlet.java

```diff
@@ -0,0 +1,120 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import com.google.sps.data.Project;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that returns one or many projects.*/
+@WebServlet("/projects")
+public class ProjectsServlet extends HttpServlet {
+
+  /** Get one or many projects. (based on whether projectId query param present)*/
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String json = "";
+    /** May be null */
+    String projectId = request.getParameter("projectId");
+
+    /** Presence of projectId flag determines whether to return one or many projects. */
+    if(projectId == null){
+        json = this.getAllProjectsJson();
+    } else {
+        try {
+            json = this.getOneProjectJson(Long.parseLong(projectId));
+        } catch (Exception badRequest) {
+            /** Client sent bad projectId */
+            response.setStatus(400);
+            response.getWriter().println("Invalid project ID.");
+            return;
+        }
+    }
+    response.setContentType("application/json");
+    response.getWriter().println(json);
+  }
+
+  /* Get JSON for a single project by id. */
+  String getOneProjectJson(long projectId) throws EntityNotFoundException {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Key projectKey = KeyFactory.createKey("Project", projectId);
+    Entity projectEntity = datastore.get(projectKey);
```

Comment:
No changes required here, but how can this fail?  What exceptions might you get?  How should you handle them?

At the moment, anything that results in an exception will result in a message to the user that their project id was invalid.

Reply:
Got it. Will explore this a bit more and patch in error codes. Have created #23 to keep track of this.